packages/unified-selection/src/unified-selection/iModel/IModelSelectionHandler.ts

```diff
@@ -0,0 +1,222 @@
+    this._activeScopeProvider = props.activeScopeProvider;
+    this._queryExecutor = props.queryExecutor;
+    this._isSuspended = false;
+    this._hasCustomCachingHiliteSetProvider = !!props.cachingHiliteSetProvider;
+
+    this._cachingHiliteSetProvider =
+      props.cachingHiliteSetProvider ??
+      createCachingHiliteSetProvider({
+        selectionStorage: this._selectionStorage,
+        iModelProvider: () => ({ queryExecutor: this._queryExecutor, metadataProvider: props.metadataProvider }),
+      });
+
+    this._hiliteSetProvider = createHiliteSetProvider({ queryExecutor: this._queryExecutor, metadataProvider: props.metadataProvider });
+    this._iModelListenerDisposeFunc = this._iModelSelection.selectionSet.onChanged.addListener(this.onIModelSelectionChanged);
+    this._unifiedSelectionListenerDisposeFunc = this._selectionStorage.selectionChangeEvent.addListener(this.onUnifiedSelectionChanged);
+
+    // stop imodel from syncing tool selection with hilited list - we want to override that behavior
+    this._iModelSelection.hilited.wantSyncWithSelectionSet = false;
+    this.applyCurrentHiliteSet({ activeSelectionAction: "clear" });
+  }
+
+  public dispose() {
+    this._cancelOngoingChanges.next();
+    this._iModelListenerDisposeFunc();
+    this._unifiedSelectionListenerDisposeFunc();
+    if (!this._hasCustomCachingHiliteSetProvider) {
+      this._cachingHiliteSetProvider.dispose();
+    }
+  }
+
+  /** Temporarily suspends tool selection synchronization until the returned `IDisposable` is disposed. */
+  public suspendIModelToolSelectionSync() {
+    const wasSuspended = this._isSuspended;
+    this._isSuspended = true;
+    return {
+      dispose: () => (this._isSuspended = wasSuspended),
+    };
+  }
+
+  private handleUnifiedSelectionChange(changeType: StorageSelectionChangeType, selectables: Selectables) {
+    if (changeType === "clear" || changeType === "replace") {
+      this.applyCurrentHiliteSet({ activeSelectionAction: "clear" });
+      return;
+    }
+
+    if (changeType === "add") {
+      from(this._hiliteSetProvider.getHiliteSet({ selectables }))
+        .pipe(takeUntil(this._cancelOngoingChanges))
+        .subscribe({
+          next: (set) => {
+            this.addHiliteSet(set);
+          },
+        });
+      return;
+    }
+
+    from(this._hiliteSetProvider.getHiliteSet({ selectables }))
+      .pipe(takeUntil(this._cancelOngoingChanges))
+      .subscribe({
+        next: (set) => {
+          this.removeHiliteSet(set);
+        },
+        complete: () => {
+          this.applyCurrentHiliteSet({ activeSelectionAction: "keep" });
+        },
+      });
+  }
+
+  private onUnifiedSelectionChanged = (args: StorageSelectionChangeEventArgs) => {
+    // iModels are only interested in top-level selection changes
+    if (args.iModelKey !== this._iModelSelection.key || args.level !== 0) {
+      return;
+    }
+
+    this._cancelOngoingChanges.next();
```

Comment:
Should we always cancel on unified selection change? Imagine the following situation:
1. User selects `X`, we start processing until "cancel".
2. User ctrl+clicks on `Y`, expecting it to be **added** to selection. However, this cancels handling of step 1.
3. Finally, only `Y` is hilited.

Reply:
Changed to not cancel when adding or removing.